### ISSUE TITLE
Enhancement request: Add token value filtering to token-store methods

### ISSUE THREAD

**Issue by platinumazure** [eslint/eslint#4563]
When looking for a particular token between two other nodes or tokens, the best available approach at the moment seems to be using `getTokensBetween` and then iterating over the tokens manually. It would be awesome to be able to pass in a token value to those methods (or similar methods) and have some of the grunt work for looping be done in token-store or SourceCode.

Below are a couple of concrete proposals. I'm not attached to all of them, or really any of them. I just wanted to show some examples of what I'm getting at to get the ball rolling.

* `getFirstToken(node)` could be overloaded to take in a value (`getFirstToken(node, value)`) which will return the first token with that value in the node.
* `getLastToken(node)` could be overloaded to take in a value (`getLastToken(node, value)`) which will return the last token with that value in the node.
* `getTokenBefore(nodeOrToken)` could be overloaded to take in a value (`getTokenBefore(nodeOrToken, value)`) which will return the last token with the given value to occur before the reference node or token (not necessarily immediately before the reference node or token).
* `getTokenAfter(nodeOrToken)` could be overloaded to take in a value (`getTokenAfter(nodeOrToken, value)`) which will return the first token with the given value to occur after the reference node or token (not necessarily immediately after the reference node or token).
* New method `getTokenBetween(node1, node2, value)` could be created, which essentially calls `getTokensBetween(node1, node2)` and returns the first token whose value matches the given value.

As you can see, all of these are themed as "get tokens before/after/between nodes, then filter by value and return single token object". So that's sort of the general concept I'm after. `getTokenBetween` in particular would allow me to revisit a recent pull request of mine to simplify it.

What do you all think?

**Comment by eslintbot** [comment]
@platinumazure Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by nzakas** [comment]
Can you show some examples where this could be used? I can see some value in providing filtering, just not sure about the implementation you're suggesting. Some concrete examples would help.

**Comment by mysticatea** [comment]
I like predicate functions (like `Array.prototype.filter`) instead of `value`.

As far as I know, this feature can be used at:

- Getting a token excludes `","`
  - https://github.com/eslint/eslint/blob/master/lib/rules/comma-dangle.js#L86
  - https://github.com/eslint/eslint/blob/master/lib/rules/object-curly-spacing.js#L193
- Getting a token of `"("`
  - https://github.com/eslint/eslint/blob/master/lib/rules/space-before-function-paren.js#L66
  - https://github.com/eslint/eslint/blob/master/lib/rules/no-spaced-func.js#L28

**Comment by platinumazure** [comment]
I'm okay with predicates too. I just like what underscore/lodash did in
some cases where they accept values that are transformed internally to
predicate functions.

In addition to examples provided by @mysticatea, what inspired this
originally from me is my recent adjustment to eqeqeq, where I need to get
"==" or "!=" tokens between the left and right side of a BinaryExpression.

**Comment by nzakas** [comment]
So what are we saying here? That you're fine with using predicate functions instead of new methods?

**Comment by platinumazure** [comment]
I'd like to have new methods (or optional parameters added to existing methods) added to token-store and source-code which will take a predicate function to allow for some filtering of what gets returned.

Nice-to-have would be a further overload which supports value predicates (e.g., `"!="` will assume that we are looking for a token with `value: "!="`), or object predicates (e.g., `{ value: "!=" }` will assume that we are looking for a token with that value). Those are just for convenience.

The basic concept remains the same: Instead of getting only the immediately previous or following token, or all tokens between two nodes, we should be able to provide more information about *exactly* what we need and token-store or source-code should be smart enough to find *that* token instead of... what it currently finds.

If you need something more concrete than that, I'm happy to refine the above to take @mysticatea's predicates into account.

**Comment by nzakas** [comment]
Got it. I think new methods are better.

**Comment by platinumazure** [comment]
Sounds good. Waiting on an accept at this point, unless you want me to refine my proposal above?

**Comment by nzakas** [comment]
Good to go. What I'm expecting is `find*` methods that correspond to the `get*` methods and accept a predicate function as an argument. Does that match what you're thinking?

**Comment by platinumazure** [comment]
Sure. I'll try to draw up a PR this week.

Good to go. What I'm expecting is find* methods that correspond to the get*
methods and accept a predicate function as an argument. Does that match
what you're thinking?

—
Reply to this email directly or view it on GitHub
<https://github.com/eslint/eslint/issues/4563#issuecomment-161150488>.

**Comment by platinumazure** [comment]
@nzakas One question, should the `find*` methods in general return arrays of tokens, or just the first matching token? Should we include `filter*` (or a `{ multiple: true }` option, or...) to allow for both behaviors?

**Comment by nzakas** [comment]
@platinumazure can you list out the methods you're thinking of so we can be sure to be on the same page?

**Comment by nzakas** [comment]
@platinumazure are you still interested in doing this?

**Comment by platinumazure** [comment]
Sorry @nzakas, I've been really busy. I concede that there has not been
nearly enough evidence this is actually needed. Could we perhaps close
this, and maybe I can reopen the issue if I get around to this in the
future? Thanks!

**Comment by nzakas** [comment]
Sounds good.

**Comment by platinumazure** [comment]
Implemented in #7936, yay!
